pkg/helm/cmd.go

```diff
@@ -0,0 +1,149 @@
+			return err
+		}
+		if !utils.IsCRD(obj) {
+			if err = corev1alpha1.AddComponentPlanLabel(obj, planName); err != nil {
+				return err
+			}
+		}
+		_, err := controllerutil.CreateOrUpdate(ctx, client, obj, func() error {
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// UnInstallByManifest remove a helm chart from the cluster
+func UnInstallByManifest(ctx context.Context, client client.Client, manifests []string) (err error) {
+	for _, manifest := range manifests {
+		obj := &unstructured.Unstructured{}
+		if err = json.Unmarshal([]byte(manifest), obj); err != nil {
+			return err
+		}
+		err = client.Delete(ctx, obj)
+		if utils.IgnoreNotFound(err) != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// UnInstallByResources remove a helm chart from the cluster
+func UnInstallByResources(ctx context.Context, client client.Client, ns string, resources []corev1alpha1.Resource) (err error) {
+	for _, resource := range resources {
+		obj := &unstructured.Unstructured{}
+		obj.SetKind(resource.Kind)
+		obj.SetName(resource.Name)
+		obj.SetAPIVersion(resource.APIVersion)
+		obj.SetNamespace(ns)
+		err = client.Delete(ctx, obj)
+		if utils.IgnoreNotFound(err) != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetManifests get helm templates
+func GetManifests(ctx context.Context, logger logr.Logger, name, namespace, chart, version, repoName, repoUrl string, set, setString, setFile, SetJSON, SetLiteral []string, skipCrd, isOCI bool) (data []string, err error) {
+	Objs, err := getHelmTemplate(ctx, logger, name, namespace, chart, version, repoName, repoUrl, set, setString, setFile, SetJSON, SetLiteral, skipCrd, isOCI)
+	if err != nil {
+		return nil, err
+	}
+	manifests := make([]*unstructured.Unstructured, 0)
+	for _, obj := range Objs {
+		if obj == nil {
+			continue
+		}
+
+		var targets []*unstructured.Unstructured
+		if obj.IsList() {
+			err = obj.EachListItem(func(object runtime.Object) error {
+				unstructuredObj, ok := object.(*unstructured.Unstructured)
+				if ok {
+					targets = append(targets, unstructuredObj)
+					return nil
+				}
+				return fmt.Errorf("resource list item has unexpected type")
+			})
+			if err != nil {
+				return nil, err
+			}
+		} else if utils.IsNullList(obj) {
+			// noop
+		} else {
+			targets = []*unstructured.Unstructured{obj}
+		}
+
+		manifests = append(manifests, targets...)
+	}
+	data = make([]string, len(manifests))
+	for i, m := range manifests {
+		manifestStr, err := json.Marshal(m.Object)
+		if err != nil {
+			return nil, err
+		}
+		data[i] = string(manifestStr)
+	}
+	return data, nil
+}
+
+func getHelmTemplate(ctx context.Context, logger logr.Logger, name, namespace, chart, version, repoName, repoUrl string, set, setString, setFile, SetJSON, SetLiteral []string, skipCrd, isOCI bool) ([]*unstructured.Unstructured, error) {
+	h := NewHelm("", isOCI)
+	out, err := h.repoAdd(ctx, repoName, repoUrl)
```

Comment:
We should do `repoAdd` in repository watcher when  `CR Repository` created, so `helm cache` will keep consistency with our `repo component list`.

Just a reminder,seems helm cli only used when get template for now. is it better to write the `template generation logic` instead of using full `helm cli`?

Reply:
> We should do `repoAdd` in repository watcher when `CR Repository` created, so `helm cache` will keep consistency with our `repo component list`.

@0xff-dev any idea?



> write the `template generation logic` instead of using full `helm cli`?

Use `helm cli` is easy to use in combination with helm plugins https://github.com/cdwv/awesome-helm#plugins, for example to use Google Cloud Storage or Amazon S3 storage to save/load the Helm charts.

Reply:
🆗  Great plugins!  let's keep helm cli for now.